ops/testing.py

```diff
@@ -242,25 +231,36 @@ def update_relation_data(self, relation_id, app_or_unit, key_values):
         :param key_values: Each key/value will be updated in the relation data.
         :return: None
         """
+        relation_name = self._backend._relation_names[relation_id]
+        relation = self._model.get_relation(relation_name, relation_id)
+        if '/' in app_or_unit:
+            entity = self._model.get_unit(app_or_unit)
+        else:
+            entity = self._model.get_app(app_or_unit)
+        rel_data = relation.data.get(entity, None)
+        if rel_data is not None:
+            # If we have read and cached this data, make sure we invalidate it
```

Comment:
Isn't the _get_ above going to read and cache it?

Reply:
_load (which calls 'relation-get') is triggered when you access an attribute (relation.data[foo]._data). 

So accessing rel_data[key] actually triggers the load, not grabbing the rel_data object.

Reply:
I probably didn't state that as well as I could.
`relation.data.get(entity)`
just returns the RelationDataContent object. That object is a LazyMapping so it doesn't actually call _load (triggering 'relation-get') until you access an attribute (either len(), iter() or __getitem__).

So it doesn't actually read relation-get just grabbing the idea-of-an-entity's data.